File: ClaudeRelayAndroid/core-protocol/src/main/kotlin/relay/protocol/ConnectionQuality.kt
package relay.protocol

/**
 * Connection quality bucket derived from median ping RTT and ping success rate.
 *
 * Ports `ConnectionQuality.swift`. The bucketing in [of] mirrors the Swift
 * `init(medianRTT:successRate:)` ordering exactly. [DISCONNECTED] is not
 * produced by [of] — it is set by transport-level logic, matching Swift where
 * the initializer never returns `.disconnected`.
 */
enum class ConnectionQuality {
    EXCELLENT,
    GOOD,
    POOR,
    VERY_POOR,
    DISCONNECTED;

    companion object {
        // RTT thresholds in seconds. 83% success = 5 of 6 pings; perfect required for excellent.
        private const val EXCELLENT_RTT = 0.1
        private const val GOOD_RTT = 0.3
        private const val POOR_RTT = 0.8
        private const val MIN_SUCCESS_RATE = 0.5
        private const val GOOD_SUCCESS_RATE = 0.83
        private const val PERFECT_SUCCESS_RATE = 1.0

        fun of(medianRttSec: Double, successRate: Double): ConnectionQuality = when {
            successRate < MIN_SUCCESS_RATE -> VERY_POOR
            medianRttSec < EXCELLENT_RTT && successRate >= PERFECT_SUCCESS_RATE -> EXCELLENT
            medianRttSec < GOOD_RTT && successRate >= GOOD_SUCCESS_RATE -> GOOD
            medianRttSec < POOR_RTT && successRate >= MIN_SUCCESS_RATE -> POOR
            else -> VERY_POOR
        }
    }
}

File: Sources/CodeRelayKit/Models/ConnectionQuality.swift
import Foundation

public enum ConnectionQuality: String, Sendable, Equatable {
    case excellent
    case good
    case poor
    case veryPoor
    case disconnected

    // RTT thresholds in seconds. 83% success = 5 of 6 pings; perfect required for excellent.
    private static let excellentRTT: TimeInterval = 0.1
    private static let goodRTT: TimeInterval = 0.3
    private static let poorRTT: TimeInterval = 0.8
    private static let minSuccessRate: Double = 0.5
    private static let goodSuccessRate: Double = 0.83
    private static let perfectSuccessRate: Double = 1.0

    public init(medianRTT: TimeInterval, successRate: Double) {
        if successRate < Self.minSuccessRate {
            self = .veryPoor
        } else if medianRTT < Self.excellentRTT && successRate >= Self.perfectSuccessRate {
            self = .excellent
        } else if medianRTT < Self.goodRTT && successRate >= Self.goodSuccessRate {
            self = .good
        } else if medianRTT < Self.poorRTT && successRate >= Self.minSuccessRate {
            self = .poor
        } else {
            self = .veryPoor
        }
    }
}
